### PR TITLE
Adding path for using CourseRunType for Seat management

### PR DESCRIPTION
**Pull request by Dillon-Dumesnil**
This adds in functionality to the course and course run
end points to use the course run type model (and through it,
the Track and SeatType models) to inform how to manage Seats
for a course run.

DISCO-1379

### CODE REVIEW

**Comment by Dillon-Dumesnil** [comment]
I think I covered the cases we need to support in my tests, but if there are any other scenarios you can think of that should be tested, I'm happy to update.

File: course_discovery/apps/api/v1/views/courses.py [hunk 1]
```diff
@@ -339,7 +339,11 @@ def update_course(self, data, partial=False):  # pylint: disable=too-many-statem
             course.entitlements = entitlements
             # If entitlements were updated, we also want to update seats
             for course_run in course.active_course_runs:
-                course_run.update_or_create_seats()
+                run_type_uuid = None
+                # DISCO-1399: Won't need the if anymore
```
Comment (mikix) [inline course_discovery/apps/api/v1/views/courses.py]: But will need the stuff inside the `if`, yah? Maybe could be clearer, but super nit.

File: course_discovery/apps/api/v1/views/courses.py [hunk 2]
```diff
@@ -339,7 +339,11 @@ def update_course(self, data, partial=False):  # pylint: disable=too-many-statem
...
+                if data.get('type'):
```
Comment (mikix) [inline course_discovery/apps/api/v1/views/courses.py]: Should this be `if course_run.type` instead? (To be more defensive around serializer being busted and to handle existing type values?)

File: course_discovery/apps/course_metadata/models.py [hunk 3]
```diff
@@ -1434,68 +1434,106 @@ def search(cls, query):
...
+        # In the CourseType flow, we only send in a single price for all of the tracks (for now),
+        # so it would be possible to accidentally set an Audit seat to have a nonzero price because
+        # there is a Verified seat with a price.
+        if seat_type == Seat.AUDIT:
             price = 0.00
 
         seat, __ = Seat.everything.update_or_create(  # pylint: disable=no-member
             course_run=self,
-            type=mode,
+            type=seat_type,
```
Comment (mikix) [inline course_discovery/apps/course_metadata/models.py:1477]: Nice.

File: course_discovery/apps/course_metadata/models.py [hunk 4]
```diff
@@ -1434,68 +1434,106 @@ def search(cls, query):
...
-                draft=True,
-                defaults={
-                    'price': 0.00,
-                    'upgrade_deadline': None,
-                },
-            )
-            seats.append(audit_seat)
+        return seat
+
+    def update_or_create_seats(self, run_type_uuid=None, price=0.00):
```
Comment (mikix) [inline course_discovery/apps/course_metadata/models.py]: Why not just `run_type=None` and take an object? Would let you simplify the `if run.type` above and just pass in `run.type`. You'd have to do a `get()` in the create call, but that's not so bad...

File: course_discovery/apps/course_metadata/models.py [hunk 5]
```diff
@@ -1434,68 +1434,106 @@ def search(cls, query):
...
+        course run exists. After an official version of the course run exists, it only supports
+        price changes or upgrades from Audit -> Verified.
+        """
+        seats = []
+        # DISCO-1399: Shouldn't need the whole if/else anymore. Can also look into if it's necessary
+        # to keep all of the helper functions (helper, validate, and upgrade deadline)
+        if run_type_uuid:
+            run_type = CourseRunType.objects.get(uuid=run_type_uuid)
+            seat_types = []
+            for track in run_type.tracks.filter(seat_type__isnull=False):
```
Comment (mikix) [inline course_discovery/apps/course_metadata/models.py]: Could also `exclude(seat_type=None)` which *might* be clearer than filtering on a null being false. But it's a toss-up.

File: course_discovery/apps/course_metadata/models.py [hunk 6]
```diff
@@ -1434,68 +1434,106 @@ def search(cls, query):
...
+                    draft=True,
+                    defaults={
+                        'price': 0.00,
+                        'upgrade_deadline': None,
+                    },
+                )
+                seats.append(audit_seat)
+            else:
+                self.seats.exclude(type=seat_type).delete()
         self.seats.set(seats)
```
Comment (mikix) [inline course_discovery/apps/course_metadata/models.py]: Does this line do anything? This is the wrong side of the relation to set yah? They are already in `self.seats` from pointing at self? Unless this is like a `refresh_from_db` but for one field?

**Review by mikix — APPROVED** [review]

File: course_discovery/apps/api/v1/views/courses.py [hunk 7]
```diff
@@ -337,9 +337,12 @@ def update_course(self, data, partial=False):  # pylint: disable=too-many-statem
                 )
         if entitlements:
             course.entitlements.set(entitlements)
-            # If entitlements were updated, we also want to update seats
-            for course_run in course.active_course_runs:
-                course_run.update_or_create_seats()
+            # DISCO-1399: This whole if can be removed. Updating or creating seats
+            # is being moved to the course run endpoint.
+            if not data.get('type'):
```
Comment (mikix) [inline course_discovery/apps/api/v1/views/courses.py]: This might be better as `if not course.type`?

**Pull request merged**